backend/adapter_utils.go

```diff
@@ -0,0 +1,154 @@
+
+func setupContext(ctx context.Context, endpoint Endpoint) context.Context {
+	ctx = WithEndpoint(ctx, endpoint)
+	ctx = propagateTenantIDIfPresent(ctx)
+
+	return ctx
+}
+
+func wrapHandler(ctx context.Context, pluginCtx PluginContext, next handlerWrapperFunc) error {
+	ctx = setupHandlerContext(ctx, pluginCtx)
+	wrapper := logWrapper(next)
+	wrapper = metricWrapper(wrapper)
+	wrapper = tracingWrapper(wrapper)
+	_, err := wrapper(ctx)
+	return err
+}
+
+func setupHandlerContext(ctx context.Context, pluginCtx PluginContext) context.Context {
+	ctx = WithGrafanaConfig(ctx, pluginCtx.GrafanaConfig)
+	ctx = WithPluginContext(ctx, pluginCtx)
+	ctx = WithUser(ctx, pluginCtx.User)
+	ctx = withContextualLogAttributes(ctx, pluginCtx)
+	ctx = WithUserAgent(ctx, pluginCtx.UserAgent)
+	return ctx
+}
+
+var pluginRequestCounter = prometheus.NewCounterVec(prometheus.CounterOpts{
+	Namespace: "grafana_plugin",
+	Name:      "request_total",
+	Help:      "The total amount of plugin requests",
+}, []string{"endpoint", "status"})
+
+var once = sync.Once{}
+
+func metricWrapper(next handlerWrapperFunc) handlerWrapperFunc {
+	once.Do(func() {
+		prometheus.MustRegister(pluginRequestCounter)
+	})
+
+	return func(ctx context.Context) (RequestStatus, error) {
+		endpoint := EndpointFromContext(ctx)
+		status, err := next(ctx)
+
+		// TODO include error/status source
+		pluginRequestCounter.WithLabelValues(endpoint.String(), status.String()).Inc()
+
+		return status, err
+	}
+}
+
+func tracingWrapper(next handlerWrapperFunc) handlerWrapperFunc {
+	return func(ctx context.Context) (RequestStatus, error) {
+		endpoint := EndpointFromContext(ctx)
+		pluginCtx := PluginConfigFromContext(ctx)
+		ctx, span := tracing.DefaultTracer().Start(ctx, fmt.Sprintf("sdk.%s", endpoint), trace.WithAttributes(
+			attribute.String("plugin_id", pluginCtx.PluginID),
+			attribute.Int64("org_id", pluginCtx.OrgID),
+		))
+		defer span.End()
+
+		if pluginCtx.DataSourceInstanceSettings != nil {
+			span.SetAttributes(
+				attribute.String("datasource_name", pluginCtx.DataSourceInstanceSettings.Name),
+				attribute.String("datasource_uid", pluginCtx.DataSourceInstanceSettings.UID),
+			)
+		}
+
+		if u := pluginCtx.User; u != nil {
+			span.SetAttributes(attribute.String("user", pluginCtx.User.Name))
+		}
+
+		status, err := next(ctx)
+
+		span.SetAttributes(
+			attribute.String("request_status", status.String()),
+		)
+
+		if err != nil {
+			return status, tracing.Error(span, err)
+		}
+
+		return status, err
+	}
+}
+
+func logWrapper(next handlerWrapperFunc) handlerWrapperFunc {
+	return func(ctx context.Context) (RequestStatus, error) {
+		start := time.Now()
+		status, err := next(ctx)
+
+		logParams := []any{
+			"status", status.String(),
+			"duration", time.Since(start),
+		}
+
+		if err != nil {
+			logParams = append(logParams, "error", err)
+		}
+
+		// TODO status source
+		// logParams = append(logParams, "statusSource", pluginrequestmeta.StatusSourceFromContext(ctx))
```

Comment:
do you want to keep this?

Reply:
Related to description
> Note 2) logs, metrics and traces currently lacks error source (status source in grafana) denoting whether an error originates from the plugin itself or from downstream service. There's some additional work needed to support this for all endpoints and would need some additional thinking to simplify for plugin authors.

I guess can target that in a follow up though. WDYT?

Reply:
yes, I just wanted to make sure you want to keep this commented code.

Reply:
Added support for error source (status source) to make it in pair with how it works in Grafana, however if a plugin returns a `backend.DownstreamError(...)` from CheckHealth or CallResource it will be recorded as a downstream error source. This should be enough for now and support the very most basic stuff.

Realized that having the `ErrorSource` in the backend package is not optimal. Tried to move the context support for error source to experimental/errorsource, but that creates a circular dependency 😢 So the current state of error status seems a bit broken to be used both internally in the SDK and in plugins. We can hopefully revisit in the future and introduce some breaking changes and clear up the problems.